provider/ec2/environ.go

```diff
@@ -66,11 +67,21 @@ var (
 	_ environs.FirewallFeatureQuerier = (*environ)(nil)
 )
 
+type ec2Client interface {
+	DescribeAvailabilityZones(*sdkec2.DescribeAvailabilityZonesInput) (*sdkec2.DescribeAvailabilityZonesOutput, error)
+	DescribeInstances(*sdkec2.DescribeInstancesInput) (*sdkec2.DescribeInstancesOutput, error)
+	DescribeInstanceTypeOfferings(*sdkec2.DescribeInstanceTypeOfferingsInput) (*sdkec2.DescribeInstanceTypeOfferingsOutput, error)
+	DescribeInstanceTypes(*sdkec2.DescribeInstanceTypesInput) (*sdkec2.DescribeInstanceTypesOutput, error)
+	DescribeSpotPriceHistory(*sdkec2.DescribeSpotPriceHistoryInput) (*sdkec2.DescribeSpotPriceHistoryOutput, error)
+}
+
 type environ struct {
 	name  string
 	cloud environscloudspec.CloudSpec
 	ec2   *ec2.EC2
```

Comment:
Add a comment that ec2.EC2 is legacy and we are migrating away from it yada yada

Reply:
Comment: sounds good, will add.

Regarding this:

```
var _ ec2Client = (ec2iface.EC2API)(nil)
```

Won't assigning the real EC2 client to an ec2Client type achieve the same thing? This happens in https://github.com/juju/juju/blob/eff28cbb84b570ad384c8d3868d3e3aad817cf42/provider/ec2/instancetypes.go#L63

Reply:
Per discussion on Mattermost, it's not any "type safer", but the `var _ ec2Client = (ec2iface.EC2API)(nil)` line makes it explicit to the reader what we're trying to implement (and gives an error message closer to the definition if you get something wrong).